app/models/vim_performance_state.rb

```diff
@@ -46,22 +51,29 @@ class VimPerformanceState < ApplicationRecord
   # => host_sockets     (derive from assoc_ids)
 
   # TODO: do a single query for the finds
+  # @param objs [Array[Object]|Object] MetricsCapture#target - all should be the same object class
+  # @param ts [Time|nil]  timestamp for capture. only used when capturing containers (default: now)
+  # @returns [Array[VimPerformanceState]|VimPerformanceState] return an array if an array was passed in 
   # NOTE: a few calls (with a single object) do use the return and expect a single result
-  def self.capture(objs)
-    ts = Time.now.utc
+  def self.capture(objs, ts = nil)
+    passed_in_an_array = objs.kind_of?(Array)
+    objs = Array(objs)
+
+    # Soft deletes can capture in the past (keep the timestamp), otherwise capture current state (timestamp = now)
+    ts = Time.now if ts.nil? || !objs.first.class.respond_to?(:active_for)
+    ts = ts.utc
```

Comment:
FYI, existing `rollup_spec.rb` fails with 

```
     NoMethodError:
       undefined method `utc' for "2010-04-14T21:00:00Z":String
     # ./app/models/vim_performance_state.rb:64:in `capture'
     # ./app/models/metric/ci_mixin/state_finders.rb:34:in `vim_performance_state_for_ts'
     # ./app/models/metric/ci_mixin/state_finders.rb:82:in `containers_from_vim_performance_state_for_ts'
     # ./app/models/metric/rollup.rb:271:in `rollup_child_metrics'
     # ./app/models/metric/rollup.rb:211:in `block in rollup_hourly'
     # ./app/models/metric/rollup.rb:211:in `each'
     # ./app/models/metric/rollup.rb:211:in `rollup_hourly'
     # ./app/models/metric/ci_mixin/rollup.rb:96:in `block (2 levels) in perf_rollup'
     # /Users/joerafaniello/.gem/ruby/3.0.6/gems/more_core_extensions-4.4.0/lib/more_core_extensions/core_ext/benchmark/realtime_store.rb:20:in `realtime_store'
     # /Users/joerafaniello/.gem/ruby/3.0.6/gems/more_core_extensions-4.4.0/lib/more_core_extensions/core_ext/benchmark/realtime_store.rb:56:in `realtime_block'
     # ./app/models/metric/ci_mixin/rollup.rb:95:in `block in perf_rollup'
     # /Users/joerafaniello/.gem/ruby/3.0.6/gems/more_core_extensions-4.4.0/lib/more_core_extensions/core_ext/benchmark/realtime_store.rb:20:in `realtime_store'
     # /Users/joerafaniello/.gem/ruby/3.0.6/gems/more_core_extensions-4.4.0/lib/more_core_extensions/core_ext/benchmark/realtime_store.rb:62:in `realtime_block'
     # ./app/models/metric/ci_mixin/rollup.rb:82:in `perf_rollup'
     # ./spec/models/metric/ci_mixin/rollup_spec.rb:439:in `block (4 levels) in <main>'
     # /Users/joerafaniello/.gem/ruby/3.0.6/gems/webmock-3.18.1/lib/webmock/rspec.rb:37:in `block (2 levels) in <main>'
```

Reply:
It's doing `container_image.perf_rollup("2010-04-14T21:00:00Z", 'hourly')`

Reply:
I went ahead and fixed a caller and added code to protect this.
This code has been moved into 22589 (and this is built upon that)